File: mpp/lab4/server/src/main/java/repository/ScoreRepository.java
package repository;

import domain.Participant;
import domain.Score;
import domain.ScoreType;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class ScoreRepository extends DatabaseRepository
        implements IScoreRepository {
    public ScoreRepository(Connection connection) throws RepositoryError {
        super(connection);
    }

    @Override
    public void createTable() throws RepositoryError {
        if (isTableCreated("Scores")) {
            return;
        }

        String command =
                "create table Scores(" +
                    "ParticipantId integer not null, " +
                    "ScoreType integer not null, " +
                    "ScoreValue integer not null, " +
                    "primary key(ParticipantId, ScoreType), " +
                    "foreign key(ParticipantId) references Participants(ParticipantId)" +
                ");";

        try {
            Statement statement = connection.createStatement();
            statement.execute(command);
            logger.info("Created table Scores");
            logger.info(command);
        } catch (SQLException e) {
            logger.error("Failed to create table Scores");
            logger.error(command);
            logger.error(e);
            throw new RepositoryError(e);
        }
    }

    @Override
    public void addScore(int participantId, ScoreType type, int score) throws RepositoryError {
        String command =
                "insert into Scores (" +
                    "ParticipantId, " +
                    "ScoreType, " +
                    "ScoreValue" +
                ") values (" +
                    "?, " +
                    "?, " +
                    "?" +
                ");";

        try {
            PreparedStatement statement = connection.prepareStatement(command);
            statement.setInt(1, participantId);
            statement.setInt(2, type.ordinal());
            statement.setInt(3, score);
            statement.executeUpdate();
            logger.info("Added score");
            logger.info(command);
        } catch (SQLException e) {
            logger.error("Failed to add score");
            logger.error(command);
            logger.error(e);
            throw new RepositoryError(e);
        }
    }

    @Override
    public Score getScore(int participantId) throws RepositoryError {
        String command =
                "select " +
                    "P.ParticipantName, " +
                    "sum(case when S.ScoreType=? then S.ScoreValue else 0 end), " +
                    "sum(case when S.ScoreType=? then S.ScoreValue else 0 end), " +
                    "sum(case when S.ScoreType=? then S.ScoreValue else 0 end) " +
                "from Participants P " +
                "left outer join Scores S on S.ParticipantId=P.ParticipantId " +
                "where P.ParticipantId=? " +
                ";";

        try {
            PreparedStatement statement = connection.prepareStatement(command);
            statement.setInt(1, ScoreType.CYCLING.ordinal());
            statement.setInt(2, ScoreType.RUNNING.ordinal());
            statement.setInt(3, ScoreType.SWIMMING.ordinal());
            statement.setInt(4, participantId);
            ResultSet results = statement.executeQuery();

            Score score;
            if (results.next()) {
                String name = results.getString(1);

                Participant participant = new Participant(participantId, name);
                score = new Score(participant);

                score.setScore(ScoreType.CYCLING, results.getInt(2));
                score.setScore(ScoreType.RUNNING, results.getInt(3));
                score.setScore(ScoreType.SWIMMING, results.getInt(4));
            } else {
                throw new RepositoryError("Failed to retrieve total score by id");
            }

            logger.info("Retrieved total score by id");
            logger.info(command);
            return score;
        } catch (RepositoryError | SQLException e) {
            logger.error("Failed to retrieve total score by id");
            logger.error(command);
            logger.error(e);
            throw new RepositoryError(e);
        }
    }

    @Override
    public void updateScore(int participantId, ScoreType type, int score) throws RepositoryError {
        String command =
                "update Scores set " +
                    "ScoreValue=? " +
                "where " +
                    "ParticipantId=? and " +
                    "ScoreType=?" +
                ";";

        try {
            PreparedStatement statement = connection.prepareStatement(command);
            statement.setInt(1, score);
            statement.setInt(2, participantId);
            statement.setInt(3, type.ordinal());
            statement.executeUpdate();
            logger.info("Updated score");
            logger.info(command);
        } catch (SQLException ex) {
            logger.error("Failed to update score");
            logger.error(command);
            logger.error(ex);
            throw new RepositoryError(ex);
        }
    }

    @Override
    public void setScore(int participantId, ScoreType type, int score) throws RepositoryError {
        try {
            addScore(participantId, type, score);
            return;
        } catch (RepositoryError ignored) {
        }

        updateScore(participantId, type, score);
    }

    @Override
    public List<Score> findScoresSortedByName() throws RepositoryError {
        String command =
                "select " +
                    "P.ParticipantId, " +
                    "P.ParticipantName, " +
                    "sum(case when S.ScoreType=? then S.ScoreValue else 0 end), " +
                    "sum(case when S.ScoreType=? then S.ScoreValue else 0 end), " +
                    "sum(case when S.ScoreType=? then S.ScoreValue else 0 end) " +
                "from Participants P " +
                "left outer join Scores S on S.ParticipantId=P.ParticipantId " +
                "group by P.ParticipantId, P.ParticipantName " +
                "order by P.ParticipantName " +
                ";";

        try {
            PreparedStatement statement = connection.prepareStatement(command);
            statement.setInt(1, ScoreType.CYCLING.ordinal());
            statement.setInt(2, ScoreType.RUNNING.ordinal());
            statement.setInt(3, ScoreType.SWIMMING.ordinal());
            ResultSet results = statement.executeQuery();
            List<Score> scores = new ArrayList<>();

            while (results.next()) {
                int id = results.getInt(1);
                String name = results.getString(2);

                Participant participant = new Participant(id, name);
                Score score = new Score(participant);

                score.setScore(ScoreType.CYCLING, results.getInt(3));
                score.setScore(ScoreType.RUNNING, results.getInt(4));
                score.setScore(ScoreType.SWIMMING, results.getInt(5));

                scores.add(score);
            }

            logger.info("Retrieved total score sorted by name");
            logger.info(command);
            return scores;
        } catch (SQLException e) {
            logger.error("Failed to retrieve total score sorted by name");
            logger.error(command);
            logger.error(e);
            throw new RepositoryError(e);
        }
    }

    @Override
    public List<Score> findScoresForTypeSortedDescending(ScoreType type) throws RepositoryError {
        String command =
                "select " +
                    "P.ParticipantId, " +
                    "P.ParticipantName, " +
                    "S.ScoreValue " +
                "from Scores S " +
                "inner join Participants P on S.ParticipantId=P.ParticipantId " +
                "where S.ScoreType=? and S.ScoreValue is not 0 " +
                "order by S.ScoreValue desc " +
                ";";

        try {
            PreparedStatement statement = connection.prepareStatement(command);
            statement.setInt(1, type.ordinal());
            ResultSet results = statement.executeQuery();
            List<Score> scores = new ArrayList<>();

            while (results.next()) {
                int id = results.getInt(1);
                String name = results.getString(2);

                Participant participant = new Participant(id, name);
                Score score = new Score(participant);

                score.setScore(type, results.getInt(3));

                scores.add(score);
            }

            logger.info("Retrieved scores for type sorted descending");
            logger.info(command);
            return scores;
        } catch (SQLException e) {
            logger.error("Failed to retrieve scores for type sorted descending");
            logger.error(command);
            logger.error(e);
            throw new RepositoryError(e);
        }
    }
}

File: mpp/lab5/repository/ScoreRepository.cs
using lab5.domain;
using System;
using System.Collections.Generic;
using System.Data.Common;
using lab5.utils;
using log4net;

namespace lab5.repository
{
    public class ScoreRepository : IScoreRepository
    {
        private static readonly ILog Logger = LogManager.GetLogger("ParticipantRepository");
        
        private DbConnection Connection { get; }

        public ScoreRepository(DbConnection connection)
        {
            Connection = connection;
            CreateTable();
        }

        private void CreateTable()
        {
            var command = Connection.CreateCommand(
                @"
                    create table if not exists Scores (
                        ParticipantId integer not null,
                        ScoreType integer not null,
                        ScoreValue integer not null,
                        primary key(ParticipantId, ScoreType),
                        foreign key(ParticipantId) references Participants(ParticipantId)
                    );
                ");
            try
            {
                command.ExecuteNonQuery();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to create table", e);
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to create table", e);

            }
            
            Logger.Info("Created table or already existing");
            Logger.Info(command.CommandText);
        }

        public void Add(int participantId, ScoreType type, int value)
        {
            var command = Connection.CreateCommand(
                @"
                    insert into Scores (
                        ParticipantId,
                        ScoreType,
                        ScoreValue
                    ) values (
                        @ParticipantId,
                        @ScoreType,
                        @ScoreValue
                    );
                ");
            command.AddParameterWithValue("@ParticipantId", participantId);
            command.AddParameterWithValue("@ScoreType", (int)type);
            command.AddParameterWithValue("@ScoreValue", value);
            try
            {
                command.ExecuteNonQuery();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to add score", e);
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to add score", e);
            }
            
            Logger.Info("Added score");
            Logger.Info(command.CommandText);
        }

        public List<Score> FindByTypeSortedDescending(ScoreType type)
        {
            var command = Connection.CreateCommand(
                @"
                    select
                        P.ParticipantId,
                        P.ParticipantName,
                        S.ScoreValue
                    from Scores S
                    inner join Participants P on S.ParticipantId=P.ParticipantId
                    where S.ScoreType=@ScoreType and S.ScoreValue is not 0
                    order by S.ScoreValue desc
                    ;
                ");
            command.AddParameterWithValue("@ScoreType", (int)type);

            DbDataReader reader;
            try
            {
                reader = command.ExecuteReader();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to find scores by type", e);
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to find scores by type", e);
            }

            Logger.Info("Found scores by type");
            Logger.Info(command.CommandText);
            
            var scores = new List<Score>();
            while (reader.Read())
            {
                var participantId = reader.GetInt32(0);
                var participantName = reader.GetString(1);
                var participant = new Participant(participantId, participantName);
                var score = new Score(participant);
                
                score.SetScore(type, reader.GetInt32(2));
                
                scores.Add(score);
            }

            return scores;
        }

        public List<Score> FindSortedByName()
        {
            var command = Connection.CreateCommand(
                @"
                    select
                        P.ParticipantId,
                        P.ParticipantName,
                        sum(case when S.ScoreType=@CyclingScoreType then S.ScoreValue else 0 end),
                        sum(case when S.ScoreType=@RunningScoreType then S.ScoreValue else 0 end),
                        sum(case when S.ScoreType=@SwimmingScoreType then S.ScoreValue else 0 end)
                    from Participants P
                    left outer join Scores S on S.ParticipantId=P.ParticipantId
                    group by P.ParticipantId, P.ParticipantName
                    order by P.ParticipantName
                    ;
                ");
            command.AddParameterWithValue("@CyclingScoreType", (int)ScoreType.Cycling);
            command.AddParameterWithValue("@RunningScoreType", (int)ScoreType.Running);
            command.AddParameterWithValue("@SwimmingScoreType", (int)ScoreType.Swimming);

            DbDataReader reader;
            try
            {
                reader = command.ExecuteReader();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to find scores", e);
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to find scores", e);
            }
            
            Logger.Info("Found scores");
            Logger.Info(command.CommandText);
            
            var scores = new List<Score>();
            while (reader.Read())
            {
                var participantId = reader.GetInt32(0);
                var participantName = reader.GetString(1);
                var participant = new Participant(participantId, participantName);
                var score = new Score(participant);
                
                score.SetScore(ScoreType.Cycling, reader.GetInt32(2));
                score.SetScore(ScoreType.Running, reader.GetInt32(3));
                score.SetScore(ScoreType.Swimming, reader.GetInt32(4));
                
                scores.Add(score);
            }

            return scores;
        }

        public Score Get(int participantId)
        {
            var command = Connection.CreateCommand(
                @"
                    select
                        P.ParticipantName,
                        sum(case when S.ScoreType=@CyclingScoreType then S.ScoreValue else 0 end),
                        sum(case when S.ScoreType=@RunningScoreType then S.ScoreValue else 0 end),
                        sum(case when S.ScoreType=@SwimmingScoreType then S.ScoreValue else 0 end)
                    from Participants P
                    left outer join Scores S on S.ParticipantId=P.ParticipantId
                    where P.ParticipantId=@ParticipantId
                    ;
                ");
            command.AddParameterWithValue("@ParticipantId", participantId);
            command.AddParameterWithValue("@CyclingScoreType", (int)ScoreType.Cycling);
            command.AddParameterWithValue("@RunningScoreType", (int)ScoreType.Running);
            command.AddParameterWithValue("@SwimmingScoreType", (int)ScoreType.Swimming);

            DbDataReader reader;
            try
            {
                reader = command.ExecuteReader();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to retrieve total score by id");
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to retrieve total score by id", e);
            }
            
            if (!reader.Read())
            {
                Logger.Error("Failed to retrieve total score by id");
                Logger.Error(command.CommandText);
                throw new RepositoryError("Failed to retrieve total score by id");
            }
            
            Logger.Info("Retrieved total score by id");
            Logger.Info(command.CommandText);
            
            var participantName = reader.GetString(0);
            var participant = new Participant(participantId, participantName);
            var score = new Score(participant);
            
            score.SetScore(ScoreType.Cycling, reader.GetInt32(1));
            score.SetScore(ScoreType.Running, reader.GetInt32(2));
            score.SetScore(ScoreType.Swimming, reader.GetInt32(3));

            return score;
        }

        public void Set(int participantId, ScoreType type, int value)
        {
            try
            {
                Add(participantId, type, value);
            }
            catch (RepositoryError)
            {
                Update(participantId, type, value);
            }
        }

        public void Update(int participantId, ScoreType type, int value)
        {
            var command = Connection.CreateCommand(
                @"
                    update Scores set
                        ScoreValue=@ScoreValue
                    where
                        ParticipantId=@ParticipantId and
                        ScoreType=@ScoreType
                    ;
                ");
            command.AddParameterWithValue("@ScoreValue", value);
            command.AddParameterWithValue("@ParticipantId", participantId);
            command.AddParameterWithValue("@ScoreType", (int)type);

            try
            {
                command.ExecuteNonQuery();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to update score");
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to update score", e);
            }
            
            Logger.Info("Updated score");
            Logger.Info(command.CommandText);
        }
    }
}
